src/Microsoft.AspNetCore.Routing/DependencyInjection/DispatcherServiceCollectionExtensions.cs

```diff
@@ -33,6 +34,16 @@ public static IServiceCollection AddDispatcher(this IServiceCollection services)
             //
             services.TryAddSingleton<MatcherFactory, TreeMatcherFactory>();
 
+            //
+            // Endpoint Selection
+            //
+            services.TryAddSingleton<EndpointSelector>();
+            services.TryAddSingleton<EndpointConstraintCache>();
```

Comment:
I'm going to restrain myself from commenting too heavily since the plan for now given the plan that you explained here.